conn.go

```diff
@@ -821,21 +747,56 @@ func (c *conn) exchangeProtoHeader(pID protoID) (stateFunc, error) {
 
 // readProtoHeader reads a protocol header packet from c.rxProto.
 func (c *conn) readProtoHeader() (protoHeader, error) {
-	var deadline <-chan time.Time
-	if c.connectTimeout != 0 {
-		deadline = time.After(c.connectTimeout)
+	// only read from the network once our buffer has been exhausted.
+	// TODO: this preserves existing behavior as some tests rely on this
+	// implementation detail (it lets you replay a stream of bytes). we
+	// might want to consider removing this and fixing the tests as the
+	// protocol doesn't actually work this way.
+	if c.rxBuf.Len() == 0 {
+		for {
+			if c.connectTimeout != 0 {
+				_ = c.net.SetReadDeadline(time.Now().Add(c.connectTimeout))
+			}
+
+			err := c.rxBuf.ReadFromOnce(c.net)
+			if err != nil {
+				return protoHeader{}, err
+			}
+
+			// read more if buf doesn't contain enough to parse the header
+			if c.rxBuf.Len() >= protoHeaderSize {
+				break
+			}
+		}
+
+		// reset outside the loop
+		if c.connectTimeout != 0 {
+			_ = c.net.SetReadDeadline(time.Time{})
+		}
 	}
-	var p protoHeader
-	select {
-	case p = <-c.rxProto:
-		return p, nil
-	case err := <-c.connErr:
-		return p, err
-	case fr := <-c.rxFrame:
-		return p, fmt.Errorf("readProtoHeader: unexpected frame %#v", fr)
-	case <-deadline:
-		return p, errors.New("amqp: timeout waiting for response")
+
+	buf, ok := c.rxBuf.Next(protoHeaderSize)
+	if !ok {
+		return protoHeader{}, errors.New("invalid protoHeader")
 	}
+	_ = buf[7]
```

Comment:
Unneeded?

Reply:
fixed (unclear why this was originally there...)

Reply:
I think this is related to https://github.com/golang/go/issues/14808.  I see similar patterns in `encoding/binary` in the standard library.  I'm going to add this back with a comment.

Reply:
Would it work equally well to do _ = buf[protoHeaderSize-1], just to show that's where the '7' is coming from?

Reply:
done